examples/step-50/step-50.cc

```diff
@@ -14,912 +14,1435 @@
-    // concerned about memory usage (and
-    // possibly the speed with which we can
-    // multiply with these matrices), we should
-    // use separate and different sparsity
-    // patterns for these two kinds of
-    // matrices.
-    for (unsigned int level = 0; level < n_levels; ++level)
-      {
-        {
-          DynamicSparsityPattern dsp(dof_handler.n_dofs(level),
-                                     dof_handler.n_dofs(level));
-          MGTools::make_sparsity_pattern(dof_handler, dsp, level);
+template <int dim, int fe_degree, typename number>
+void LaplaceOperator<dim, fe_degree, number>::evaluate_coefficient(
+  const Coefficient<dim> &coefficient_function)
+{
+  const unsigned int n_cells = this->data->n_macro_cells();
+  FEEvaluation<dim, fe_degree, fe_degree + 1, 1, number> phi(*this->data);
+
+  coefficient.reinit(TableIndices<1>(n_cells));
+  for (unsigned int cell = 0; cell < n_cells; ++cell)
+    {
+      phi.reinit(cell);
+
+      VectorizedArray<number> averaged_value(0);
+      for (unsigned int q = 0; q < phi.n_q_points; ++q)
+        averaged_value += coefficient_function.value(phi.quadrature_point(q));
+      averaged_value /= phi.n_q_points;
+
+      coefficient(cell) = averaged_value;
+    }
+}
+
+
+template <int dim, int fe_degree, typename number>
+Table<1, VectorizedArray<number>>
+LaplaceOperator<dim, fe_degree, number>::get_coefficient_table()
+{
+  return coefficient;
+}
 
-          mg_matrices[level].reinit(dof_handler.locally_owned_mg_dofs(level),
-                                    dof_handler.locally_owned_mg_dofs(level),
-                                    dsp,
-                                    mpi_communicator,
-                                    true);
-        }
 
+template <int dim, int fe_degree, typename number>
+void LaplaceOperator<dim, fe_degree, number>::local_apply(
+  const MatrixFree<dim, number> &                   data,
+  LinearAlgebra::distributed::Vector<number> &      dst,
+  const LinearAlgebra::distributed::Vector<number> &src,
+  const std::pair<unsigned int, unsigned int> &     cell_range) const
+{
+  FEEvaluation<dim, fe_degree, fe_degree + 1, 1, number> phi(data);
+
+  for (unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+    {
+      AssertDimension(coefficient.size(0), data.n_macro_cells());
+
+      phi.reinit(cell);
+      phi.read_dof_values(src);
+      phi.evaluate(false, true);
+      for (unsigned int q = 0; q < phi.n_q_points; ++q)
+        phi.submit_gradient(coefficient(cell) * phi.get_gradient(q), q);
+      phi.integrate(false, true);
+      phi.distribute_local_to_global(dst);
+    }
+}
+
+
+template <int dim, int fe_degree, typename number>
+void LaplaceOperator<dim, fe_degree, number>::apply_add(
+  LinearAlgebra::distributed::Vector<number> &      dst,
+  const LinearAlgebra::distributed::Vector<number> &src) const
+{
+  this->data->cell_loop(&LaplaceOperator::local_apply, this, dst, src);
```

Comment:
Given that the other tutorials on matrix-free methods use the function pointer approach, I would leave the decision to the discretion of the authors. It would need at least some short explanation - but I must admit I would like a lambda example somewhere in the tutorial, but I am not insisting here.

There are arguments both in favor of lambdas and named functions. The former is very compact and would be a good option here (especially in C++14 or newer where you can use the `auto` keyword for arguments as you do). On the other hand, named functions for larger code blocks make the code easier to read and provide some structure that I struggle to see in lambdas. And at least with some of the tools we have today (gdb, valgrind), I find it more disturbing to attach to lambdas.

Reply:
I have no preference either way. @tjhei?